File: haskell/Menu.hs
module Menu (printMenu, chooseOption, toContinue, backToMenu) where

import System.Exit (exitSuccess)

-- EXIBE O MENU
printMenu :: IO ()
printMenu = do
    putStrLn "\n###### COMPUTING MAZE ######"
    putStrLn "1 - INICIAR O JOGO"
    putStrLn "2 - TUTORIAL"
    putStrLn "3 - SAIR DO JOGO"
    option <- readLn :: IO Int
    chooseOption option

-- SELECIONA A OPÇÃO DO USUÁRIO
chooseOption :: Int -> IO ()
chooseOption 1 = startGame
chooseOption 2 = openTutorial
chooseOption 3 = do
    putStrLn "\n###### COMPUTING MAZE ######\n"
    exitSuccess
chooseOption invalidOption = do
  putStrLn "\n### Opção inválida. Tente novamente. ###"
  printMenu

-- CONTINUA O FLUXO DE EXECUÇÃO
toContinue :: IO ()
toContinue = do
    putStrLn "\n### Pressione ENTER para continuar. ###"
    line <- getLine :: IO String
    putStrLn "\n"

-- RETORNA AO MENU DE OPÇÕES
backToMenu :: IO ()
backToMenu = do
    putStrLn "\n### Pressione ENTER para voltar ao MENU PRINCIPAL. ###"
    line <- getLine :: IO String
    printMenu

-- INICIA O JOGO
startGame :: IO ()
startGame = do
    putStrLn "\n### Escrever algo aqui. ###" -- EDITAR MENSAGEM POSTERIORMENTE
    toContinue

-- ABRE O TUTORIAL
openTutorial :: IO ()
openTutorial = do
    -- EDITAR MENSAGENS POSTERIORMENTE
    putStrLn "\nComputing Maze é um jogo 2D de simulação de labirintos "
    putStrLn "em um contexto do curso de Ciência da Computação."
    putStrLn "Os labirintos são gerados automaticamente de forma aleatória."
    putStrLn "A cada fase nova, você vai encontrar um novo labirinto e"
    putStrLn "um enigma relacionado a um período específico do fluxograma do curso."
    putStrLn "Seja cauteloso, o seu objetivo final é chegar a conclusão do curso"
    putStrLn "e conseguir sair desses labirintos com sucesso."
    toContinue
    -- EDITAR MENSAGENS POSTERIORMENTE
    putStrLn "\nOs enigmas que você vai desvendar são a chave"
    putStrLn "para conseguir sair definitivamente de cada labirinto."
    putStrLn "Mas será que você consegue descobrir os mistérios de FMCC?"
    putStrLn "Ou será que você vai conseguir solucionar a temida"
    putStrLn "Teoria da Computação? Você acha mesmo ser capaz de"
    putStrLn "solucionar os quebra-cabeças de ATAL e Compiladores?"
    putStrLn "Vamos descobrir até onde você consegue chegar nessa jornada."
    toContinue
    -- EDITAR MENSAGEM POSTERIORMENTE
    putStrLn "\n ESCREVER INSTRUÇÕES AQUI. "
    
    backToMenu

File: prolog/final/menu.pl
% menu.pl
:- dynamic choice/1.
:- consult('main.pl').
:- consult('maze.pl').
:- consult('enigmas.pl').

main_menu :-
    write('Escolha uma opção:'), nl,
    write('1. Jogar (pressione 1)'), nl,
    write('2. Tutorial (pressione 2)'), nl,
    write('3. Sair (pressione 3)'), nl,
    read_choice(Choice),
    handle_choice(Choice).

% Função para ler a escolha pressionando uma tecla
read_choice(Choice) :-
    get_single_char(Char),
    Char \= 10,  % Ignora a tecla Enter
    ( Char =:= 49 -> Choice = 1;  % '1'
      Char =:= 50 -> Choice = 2;  % '2'
      Char =:= 51 -> Choice = 3;  % '3'
      write('Escolha inválida! Tente novamente.'), nl,
      read_choice(Choice)
    ).

handle_choice(1) :- 
    iniciar_jogo,  % Chama a função para iniciar o jogo
    main_menu.  % Retorna ao menu após o jogo
handle_choice(2) :- 
    writeln('Welcome to the Tutorial:'),
    writeln('Computing Maze é um jogo 2D para simular labirintos'),
    writeln('no contexto de um curso de Ciência da Computação. Em cada novo nível,'),
    writeln('você encontrará um novo labirinto e um enigma relacionado a um período específico'),
    writeln('do currículo do curso. Tenha cautela, seu objetivo final'),
    writeln('é completar o curso e escapar com sucesso destes labirintos.'),
    writeln('Os enigmas que você resolver são a chave para escapar permanentemente'),
    writeln('de cada labirinto. Mas você consegue desvendar os mistérios da temida'),
    writeln('Teoria da Computação? Vamos descobrir até onde você pode ir nesta'),
    writeln('jornada. Para mover o jogador, você pode usar os métodos padrão'),
    writeln('utilizando as teclas W A S D.'),
    main_menu.  % Retorna ao menu após o tutorial
handle_choice(3) :- 
    write('Saindo do jogo. Até mais!'), nl, halt.

% Função para iniciar o jogo
iniciar_jogo :-
    write('Bem-vindo ao Computing Maze - Última Fase!'), nl,
    labirintos(LabList),  % Carrega a lista de labirintos
    jogar_labirintos(LabList).  % Joga apenas o último labirinto

